engine/minskyTensorOps.cc

```diff
@@ -315,26 +320,23 @@ namespace minsky
         if (arg1->hypercube().dims()[arg1->rank()-1]!=arg2->hypercube().dims()[0] && (arg1->rank()<arg2->rank()))  // do not allow products of the type nx1 dot nxm
           throw std::runtime_error("inner dimensions of tensors do not match");
         
-        // shape of output tensor should match outer dimensions of the two rank<2 tensors originally multiplied by one another     
-        if (arg1->rank()>1 && arg2->rank()>1) {
-          vector<XVector>&& outerXVs{arg1->hypercube().xvectors[0],arg2->hypercube().xvectors[arg2->rank()-1]};
-          cachedResult.hypercube(outerXVs);
-        }
-        else if (arg1->rank()>1 && arg2->rank()<2) { //case mxn dot nx1
-          vector<XVector>&& outerXVs{arg2->hypercube().xvectors[0]};
-          cachedResult.hypercube(outerXVs);
-        }
-        else {
-			//cachedResult.hypercube({});  // causes segfault: 0x00005555556d3926 in boost::any::~any (this=0x55555723f850, __in_chrg=<optimized out>) at /usr/local/include/boost/any.hpp:79 delete content; 
-			
-			// also causes the above segfault
-            auto hc=arg2->hypercube();
-            if (rank()==0) return;
-            
-            auto& xv=hc.xvectors[0];
-            xv.erase(xv.end()-xv.size()-2, xv.end());
-            cachedResult.hypercube(move(hc));
-		}
+        auto xv1=arg1->hypercube().xvectors, xv2=arg2->hypercube().xvectors;
+        Hypercube hc;
+        hc.xvectors.insert(hc.xvectors.begin(), xv1.begin(), xv1.end()-1);
+        hc.xvectors.insert(hc.xvectors.begin(), xv2.begin()+1, xv2.end());
+        cachedResult.hypercube(move(hc));
+        
+        auto idx1=arg1->index(), idx2=arg2->index();
+        set<size_t> newIdx;        
+        for (auto& i: idx1)
+          for (auto& j: idx2)
+          {
+            // strip of any indices outside the output range
+            auto t=ssize_t(j)-ssize_t(i);
+            if (t>=0 && t<ssize_t(size()))  newIdx.insert(t);
```

Comment:
This doesn't make sense

Reply:
The index() members of both arg1 and arg2 are empty in setArguments(), so I don't think I should be calculating it here, but rather in computeTensor()? In the dense case, the index is filled with the correct indices of the finite elements in the final matrix, but the index's size exceeds the number of finite elements. So, I guess the index() would be correct if it appeared filled up to only the last non-blank cell? (see attached screenshot
![Index](https://user-images.githubusercontent.com/46541934/94371472-20137400-00f7-11eb-958c-2ace3e7ddfb7.png)
)